### ISSUE TITLE
Interpreting abundance

### ISSUE THREAD

**Issue by mytluo** [dib-lab/sourmash#549]
Hi @ctb,

I ran `--track-abundance` with `sourmash gather`, and wasn't sure how to interpret the values. Should I be looking at `average_abund`?

On another note, for `--output-unassigned`, is it correct to interpret that file as the list of hashes that were unclassified for the "mins" array? If I took the length of that array and added it to the sum of `average_abund` (or the correct column), could I calculate the % abundance from that sum? Apologies if this is completely off the mark!

**Comment by luizirber** [comment]
so, `num` is a bit misleading because it's a value we use internally in sourmash to control between regular minhashes (like mash) or scaled minhashes. In regular minhashes `len(mins) == num`, but because you calculated your signatures with `--scaled` you will need to check `len(mins)` to get how many hashes were collected.

**Comment by mytluo** [comment]
@luizirber, thank you! that makes sense... I got 107323 from `len(mins)`, but when I sum up the entire `count` column from `lca summarize`, it only adds up to 67072... I would also think this is the wrong sum in general as some of the counts are redundant -- counts from _E. coli_ would also be in _Escherichia_, up to the _Bacteria_ superkingdom level. But that would lower the total even more, which is strange, given 93.7% of the query was hit. Would you be able to point me in the right direction? I apologize for all the questions, I know I've asked a lot ^^;;

**Comment by mytluo** [comment]
I realized I forgot the `--scaled` ^^;; last question (hopefully) -- in the LCA counts, would the count with no superkingdom assigned be considered unclassified reads?

**Comment by ctb** [comment]
While I try to find time to write more docs, please also see my STAMPS 2018 presentation at https://osf.io/8d56n/ -- thx!

**Comment by Alok-Shekhar** [comment]
@mytluo , @ctb 
I am not getting how to interpret souurmash gather output especially  f_unique_weighted, average_abund, median_abund, std_abund columns values, how to calculate %abundance value from the run, will you please explain.  

2. how to interpret --scaled option although its explained but am not getting correctly, will you please elaborate it for better understanding.  Thanks A lot!

**Comment by ctb** [comment]
hi @Alok-Shekhar thank you for your questions!

## scaled

 re scaled, have you seen the explanations in [Using sourmash: a practical guide](https://sourmash.readthedocs.io/en/latest/using-sourmash-a-guide.html#what-resolution-should-my-signatures-be-how-should-i-compute-them) and [modulo hash details](https://sourmash.readthedocs.io/en/latest/api-example.html#sourmash-minhash-objects-and-manipulations)?

##

re abundance, what specific question(s) do you want to answer?

here's an example (from the test code) that might help:

In the `tests/test-data/` directory, I run:
```
% sourmash gather gather-abund/reads-s10x10-s11.sig gather-abund/genome-s{10,11,12}.fa.gz.sig
== This is sourmash version 2.0.0a12.dev48+ga92289b. ==
== Please cite Brown and Irber (2016), doi:10.21105/joss.00027. ==

select query k=21 automatically.
loaded query: 1-1... (k=21, DNA)
loaded 3 signatures.


overlap     p_query p_match avg_abund
---------   ------- ------- ---------
0.5 Mbp       91.0%  100.0%      14.5    tests/test-data/genome-s10.fa.gz
376.0 kbp      9.0%   80.0%       1.9    tests/test-data/genome-s11.fa.gz

found 2 matches total;
the recovered matches hit 100.0% of the query
```

here, the query reads were constructed to have 10 times as many reads from s10 as s11. And so that's what you see with p_query (91% of the query is genome s10, 9% of the query is genome s11). In terms of the average abundance (`avg_abund`), what you see is approximately the same relative relationship (10x as much in s10 as in s11).

For this example, there is no difference between p_query and avg_abund, because the s10 and s11 genomes are the same size. However, were s10 twice as big as s11, then the p_query should change while avg_abund should not.

**Comment by Alok-Shekhar** [comment]
Thanks! @ctb for the explanation
Am I right, please let me know.
**f_unique_weighted columns** values are considered as **%abundance value**?
Thanks a lot!

**Comment by ctb** [comment]
@luizirber is doing a bunch of work on this for his thesis, so we can finally lay this to rest with benchmarks and (lots of) details. soon!

**Comment by cesarcardonau** [comment]
Hi @ctb @luizirber, since it has been a few months, I was wondering if you guys have any more details/documentation for interpreting abundances as mentioned above.  

I am struggling a bit myself, picking up these differences. For instance, how is the abundance weighing for `f_unique_weighted` work,  if it seems that doesn't require that the signature is pre-computed with the `--track-abundance` flag. So how is it able to do it? or what is different from this abundance?

Thanks!

**Comment by ctb** [comment]
hi @cesarcardonau et al., let me update you all a bit on some documentation and testing updates!

first, our [classifying signatures](https://sourmash.readthedocs.io/en/latest/classifying-signatures.html) documentation has been improved. In particular the [abundance weighting](https://sourmash.readthedocs.io/en/latest/classifying-signatures.html#abundance-weighting) section is better, and the gather algorithm discussion at the bottom of the page should be informative.

Second, we've [updated lca summarize to respect abundance info](https://github.com/dib-lab/sourmash/pull/1022) and elaborated [our documentation](https://sourmash.readthedocs.io/en/latest/command-line.html#sourmash-lca-summarize).

I don't think either of these answers the questions above fully, but they might provide context.

On to the meat of the question --

**Comment by ctb** [comment]
There's a lot of questions in this issue, so let me try to answer some of them first and then iterate as people ask for clarification!

The gather-abund tests added in #886 are key See [this diff](https://github.com/dib-lab/sourmash/pull/886/files#diff-5f3c562418acf29ea07e084b1d61594c), but I interpret the important bits below:

# sourmash gather and abundances

tl;dr Below is a discussion of a synthetic set of test cases using three randomly generated (fake) genomes of the same size, with two even read data set abundances of 2x each, and a third read data set with 20x.

## Data set prep

First, we make some synthetic data sets:

* r1.fa with 2x coverage of genome s10
* r2.fa with 10x coverage of genome s10.
* r3.fa with 2x coverage of genome s11.

then we make signature s10-s11 with r1 and r3, i.e. 1:1 abundance, and make signature s10x10-s11 with r2 and r3, i.e. 10:1 abundance.

## test_gather_abund_1_1 in test_sourmash.py

When we project r1+r3, 1:1 abundance, onto s10, s11, and s12 genomes with gather:

```
sourmash gather r1+r3 genome-s10.sig genome-s11.sig genome-s12.sig
```

we get:

```
overlap     p_query p_match avg_abund
---------   ------- ------- ---------
394.0 kbp     49.6%   78.5%       1.8    genome-s10.fa.gz
376.0 kbp     50.4%   80.0%       1.9    genome-s11.fa.gz
```

* approximately 50% of each query matching (first column, p_query)
* approximately 80% of subject genome's contents being matched (this is due to the low coverage of 2 used to build queries)
* approximately 2.0 abundance (third column, avg_abund)
* no match to genome s12.

## test_gather_abund_10_1

When we project r2+r3, 10:1 abundance, onto s10, s11, and s12 genomes with gather:

```
sourmash gather r2+r3 genome-s10.sig genome-s11.sig genome-s12.sig
```

we get:

```
overlap     p_query p_match avg_abund
---------   ------- ------- ---------
0.5 Mbp       91.0%  100.0%      14.5    tests/test-data/genome-s10.fa.gz
376.0 kbp      9.0%   80.0%       1.9    tests/test-data/genome-s11.fa.gz
```

* approximately 91% of s10 matching
* approximately 9% of s11 matching
* approximately 100% of the high coverage genome being matched, with only 80% of the low coverage genome
* approximately 2.0 abundance (third column, avg_abund) for s11, and (very) approximately 20x abundance for genome s10.

The cause of the poor approximation for genome s10 is unclear; it could be due to low coverage, or small genome size, or something else. More investigation needed.

**Comment by ctb** [comment]
I think the above should answer @mytluo first question.

answering @mytluo second question,
>for `--output-unassigned`, is it correct to interpret that file as the list of hashes that were unclassified for the "mins" array? If I took the length of that array and added it to the sum of `average_abund` (or the correct column), could I calculate the % abundance from that sum? Apologies if this is completely off the mark!

Yes, that file is the list of hashes that were not classified / assigned to a genome. The percentages and so on referred to elsewhere in this issue should take that all into account, however, so you don't need to explicitly account for them yourself!

**Comment by ctb** [comment]
more @mytluo answers -

> I was pointed to `f_unique_weighted` as a possible equivalence to relative abundance -- the summed fraction seems to match the query coverage, so that seems right, but I just wanted to confirm before moving forward :)

Yeah, in partial response to this issue, I put in some tests to make sure I understood my own code :).

Per [this line](https://github.com/dib-lab/sourmash/pull/886/files#diff-5f3c562418acf29ea07e084b1d61594cR3099), `f_unique_weighted` is the abundance-weighted fraction of k-mers uniquely assigned to that gather match. See `f_unique_to_query` discussion [here](https://sourmash.readthedocs.io/en/latest/classifying-signatures.html#appendix-how-sourmash-gather-works), which we then adjust to be weighted by the abundances.

**Comment by ctb** [comment]
the answer to @Alok-Shekhar question,

> [how can we get] Abundance info just like we get from kraken run from "% of fragments covered " column?

is in #1022 and in the [latest docs for lca summarize](https://sourmash.readthedocs.io/en/latest/command-line.html#sourmash-lca-summarize)

**Comment by ctb** [comment]
ok I think I'm done for now, thanks for all the questions and please ask more!

**Comment by cesarcardonau** [comment]
This is very helpful, thanks so much @ctb

**Comment by ctb** [comment]
thank you for sticking with & please do ask for clarification :).  we have an
increasing number of good synthetic and real data sets that we can use to
explore and demonstrate this stuff.

**Comment by DivyaRavichandar** [comment]
@ctb This is very helpful in understanding `f_unique_weighted`. Would it be reasonable to go to from `f_unique_weighted` to an approximation of reads annotated to a strain by `f_unique_weighted*read_depth_in_sample`?

**Comment by ctb** [comment]
my naive thinking is that `f_unique_weighted * number of reads` should be what you're looking for. but I haven't thought that through :)

**Comment by ctb** [comment]
closing for now - moved final unanswered question to #1079 for visibility.
